pkg/build/api/v1/types.go

```diff
@@ -402,6 +408,19 @@ type SourceBuildStrategy struct {
 	ForcePull bool `json:"forcePull,omitempty"`
 }
 
+// JenkinsPipelineBuildStrategy holds parameters specific to a Jenkins Pipeline build.
+type JenkinsPipelineBuildStrategy struct {
+	// JenkinsfilePath is the path of the Jenkinsfile that will be used to configure the pipeline
+	// relative to the root of the source repository.
+	JenkinsfilePath string `json:"jenkinsfilePath,omitempty"`
+
+	// Jenkinsfile is the raw contents of a Jenkinsfile containing the pipeline definition.
+	Jenkinsfile string `json:"jenkinsfile,omitempty"`
+
+	// Env contains additional environment variables you want to pass to the Jenkins Pipeline.
+	Env []kapi.EnvVar `json:"env,omitempty"`
```

Comment:
In general all the steps to build & test are definitely in the Jenkinsfile.

To deal with environments being on different clusters; or not even having access to environments (e.g. production), I was hoping we could store the small amount of environment metadata inside the BuildConfig which the Jenkinsfile can then use (either via env vars or loading the BuildConfig). 

The main reason is then that we can do nice tooling for editing of the environment metadata (e.g. allowing folks to reuse existing namespaces using auto completion etc). Its gonna be very common for folks to want to pick which namespaces/projects are going to be used for environments (e.g. using one namespace for staging of all apps in a project; or having one namespace per app etc).

Am thinking something simple along these lines:
```
environments:
  - name: Test
    namespace: myproject-test
  - name: Staging
    namespace: myproject-staging
  - name: Production
    namespace: myproject-prod
    cluster: something.com:8443

or for the 'cannot access environment' case:

  - name: Production
    git:
       uri: http://acme.com/myproject-production-resources.git
```

This also helps us have reusable pipelines that need less editing by users; since they can have a nice little UI to define which physical OpenShift / AEP namespaces and clusters (or git repos) are used to deploy to each environment.